lib/mozart_fetcher/decoder.ex

```diff
@@ -2,62 +2,46 @@ defmodule MozartFetcher.Decoder do
   alias MozartFetcher.{Envelope, Config}
 
   def decode_envelope(data, struct) do
-    with {:ok, decoded} <- Jason.decode(data, keys: :atoms),
-         {:ok, struct} <- to_struct(decoded, struct) do
-      {:ok, struct}
-    else
-      _ -> {:error}
+    case Jason.decode(data, keys: :atoms) do
+      {:ok, decoded} ->
+        to_struct(decoded, struct)
+
+      {:error, error = %Jason.DecodeError{}} ->
+        Stump.log(:error, "Envelope decode error: #{error.data}")
+        {:error}
     end
   end
 
-  # This could be more generic if we add a list_name then Enum.map(data[list_name]) instead of specifically for components here
   def decode_config(data, struct) do
     case Jason.decode(data, keys: :atoms) do
-      {:ok, data} ->
-        {:ok,
-         Enum.map(data[:components], fn map ->
-           case to_struct(map, struct) do
-             {:error, _} -> {:error}
-             {:ok, struct} -> struct
-           end
-         end)}
-
-      {:error, _} ->
+      {:ok, decoded} ->
+        {:ok, components_to_struct(decoded[:components], struct)}
+
+      {:error, error = %Jason.DecodeError{}} ->
+        Stump.log(:error, "Config decode error: #{error.data}")
         {:error}
     end
   end
 
-  def to_struct(map, struct = %Config{}) do
-    case [:endpoint, :format, :id, :must_succeed] |> Enum.all?(&Map.has_key?(map, &1)) do
-      true ->
-        {:ok, struct(struct, map)}
-
-      false ->
-        ExMetrics.increment("error.components.decode")
-
-        Stump.log(:error, %{
-          message: "Map contains Invalid keys cannot convert to Config",
-          map: map
-        })
-
-        {:error, "Failed to validate keys in struct"}
-    end
+  defp components_to_struct(components, struct) do
+    Enum.map(components, fn map ->
+      struct(struct, map)
+    end)
   end
 
-  def to_struct(map, struct = %Envelope{}) do
-    case [:head, :bodyInline, :bodyLast] |> Enum.all?(&Map.has_key?(map, &1)) do
-      true ->
-        {:ok, struct(struct, map)}
-
-      false ->
-        ExMetrics.increment("error.components.decode")
+  defp to_struct(map, struct = %Envelope{}) do
+    try do
+      {:ok, struct!(struct, map)}
+    rescue
```

Comment:
as function is only used for the `Envelope` struct, I think this try and rescue could be replaced by checking the keys instead? I think that'd be more Elixir like.

Also,[ I believe this `case` is now redundant](https://github.com/bbc/mozart-fetcher/blob/master/lib/mozart_fetcher/envelope.ex#L8) because we will always return `{:ok, %Envelope{}}`. And therefore i[ncorrectly increment this metric ](https://github.com/bbc/mozart-fetcher/blob/master/lib/mozart_fetcher/envelope.ex#L10)

Reply:
I have updated the codebase so that we only log specifically when we are close to the specific reasoning in the decoder and have left it so only the Envelope module will do the metricing otherwise we would end up metricing multiple times as you have suggested.

I think im going to leave the key error as I think its a bit clearer, I think it has a bit too much logic with the key checking unless someone has a better suggestion than this to check them:
`case [:head, :bodyInline, :bodyLast] |> Enum.all?(&Map.has_key?(map, &1)) do`

Reply:
Also have changed the decoder to return `{:error}` to keep things consistent and so that it is the Envelope module that returns the empty %Envelope{} on error.